generic-contracts/src/main/java/com/scalar/dl/genericcontracts/table/v1_0_0/Select.java

```diff
@@ -53,18 +160,249 @@ private JsonNode project(JsonNode records, JsonNode projections) {
+    ListMultimap<String, JsonNode> conditionsMap = ArrayListMultimap.create();
+    conditions.forEach(
+        condition -> {
+          String column = condition.get(Constants.CONDITION_COLUMN).asText();
+          if (isColumnNameWithTableReference(column)) {
+            String tableReference = getTableReference(column);
+            String columnName = getColumnName(column);
+            ObjectNode newCondition = getObjectMapper().createObjectNode();
+            newCondition.put(Constants.CONDITION_COLUMN, columnName);
+            newCondition.set(Constants.CONDITION_VALUE, condition.get(Constants.CONDITION_VALUE));
+            newCondition.set(
+                Constants.CONDITION_OPERATOR, condition.get(Constants.CONDITION_OPERATOR));
+            conditionsMap.put(tableReference, newCondition);
+          } else {
+            conditionsMap.put(defaultTableReference, condition);
+          }
+        });
+    return conditionsMap;
+  }
+
+  private JsonNode prepareScanArguments(
+      String tableName, String tableReference, ListMultimap<String, JsonNode> conditionsMap) {
+    return prepareScanArguments(tableName, tableReference, conditionsMap, null);
+  }
+
+  private JsonNode prepareScanArguments(
+      String tableName,
+      String tableReference,
+      ListMultimap<String, JsonNode> conditionsMap,
+      JsonNode joinCondition) {
+    ArrayNode conditionArray = getObjectMapper().createArrayNode();
+    if (joinCondition != null) {
+      conditionArray.add(joinCondition);
+    }
+    conditionsMap.get(tableReference).forEach(conditionArray::add);
+    return getObjectMapper()
+        .createObjectNode()
+        .put(Constants.QUERY_TABLE, tableName)
+        .set(Constants.QUERY_CONDITIONS, conditionArray);
+  }
+
+  private JsonNode getTableMetadata(Ledger<JsonNode> ledger, String tableName) {
+    String tableAssetId = getAssetIdForTable(tableName);
+    return ledger
+        .get(tableAssetId)
+        .orElseThrow(() -> new ContractContextException(Constants.TABLE_NOT_EXIST))
+        .data();
+  }
+
+  private String addTableReference(String tableReference, String columnName) {
+    return tableReference + Constants.COLUMN_SEPARATOR + columnName;
+  }
+
+  private JsonNode addTableReferenceForColumns(JsonNode record, String tableReference) {
+    ObjectNode renamed = getObjectMapper().createObjectNode();
+    Iterator<Entry<String, JsonNode>> columns = record.fields();
+
+    while (columns.hasNext()) {
+      Map.Entry<String, JsonNode> column = columns.next();
+      renamed.set(addTableReference(tableReference, column.getKey()), column.getValue());
+    }
+
+    return renamed;
+  }
+
+  private Set<String> getJoinTableReferences(JsonNode joins) {
+    Set<String> tables = new HashSet<>();
+
+    for (JsonNode join : joins) {
+      JsonNode table = join.get(Constants.JOIN_TABLE);
+      tables.add(getTableReference(table));
+    }
+
+    return tables;
+  }
+
+  private boolean isColumnName(String columnName) {
```

Comment:
It depends, but I think creating `Pattern` (compiled pattern) might be faster if we need to call this several times.
Ditto for the other regular expressions.

Reply:
Right. Since it's only used in the validation (not in the join loops), I skipped it. But, as you said, it would be better. Will do, thanks!

Reply:
Fixed in 0d8d6c9. PTAL again!